### PR TITLE
Routing

### PR DESCRIPTION
**Pull request by gnomex909**
Added skeleton of the routing of our website

### CODE REVIEW

File: frontend/Components/Footer.tsx [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import * as React from 'react';
```
Comment (EvidentlyCube) [inline frontend/Components/Footer.tsx]: Unrealted to this file but:
In `.eslintrc` change `@typescript-eslint/explicit-function-return-type` to `off`. React has a lot of types that are implied by the convention and it's just a waste of everything to be explicit about them in my opinion.

Comment (EvidentlyCube) [reply frontend/Components/Footer.tsx]: Also add this:
```
'react/prop-types': ['off'],
```

Maybe here:
image

Prop types are a built-in mechanism in React for runtime checking of props which we don't need because we have compile-time type checking with TypeScript.

File: frontend/Components/Games.tsx [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import * as React from 'react';
+export interface GameProps {
+	game: string;
+}
+
+export const GamesCheck = (props: GameProps): JSX.Element => <>
```
Comment (EvidentlyCube) [inline frontend/Components/Games.tsx]: 1. Because we're dropping explicit types declarations from linter you can remove the return type on the function.

2. But there is one thing to change:
```
export const GamesCheck: React.FC<GameProps> = (props) => <>
```
`React.FC` is the type for a functional component, and having `GameProps` in the definition of the constant rather than the arrow function makes it slightly more consistent - generally in TS you avoid defining types in arrow functions to make them simpler to parse.

File: frontend/Components/Header.tsx [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import * as React from 'react';
+import {
+	Link,
+} from 'react-router-dom';
+export interface HeaderParams{
```
Comment (EvidentlyCube) [inline frontend/Components/Header.tsx]: It's [`Props` not `Params`](https://reactjs.org/docs/components-and-props.html) :)

File: frontend/Components/Header.tsx [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+	pageName: string;
+}
+export class MainHeader extends React.Component<HeaderParams> {
+	private readonly pageName: string;
+
+	constructor(params: HeaderParams) {
+		super(params);
+		this.pageName = params.pageName;
```
Comment (EvidentlyCube) [inline frontend/Components/Header.tsx]: In React you generally avoid having fields in component classes - you use [state](https://reactjs.org/docs/state-and-lifecycle.html) instead. The benefit of state is that any change to state will automatically trigger a rerender, wherein if you use fields you'll have to... actually there's probably no decent way to trigger a rerender manually.

Some pieces of wisdom:
1. In constructor you can set state simply by `this.state = {pageName: params.pageName}`
2. You read state simply by using `this.state`
2. Any subsequent changes to state have to be done via `this.setState()`.
3. `setState()` **IS ASYNCHRONOUS**!!!!! If you call `setState()` and on the next line try to read state it'll still have the old values.
4. Changes to state trigger rerenders of the component
5. If you set the state from props in constructor, when the props change the change won't propagate to state automatically again. React tries to keep the instances of the components alive. If for some reason you need to update the state if the props passed to a component change, you need [getDerivedStateFromProps](https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops). But if you ever feel you need to ask me for feedback first before you go too deep into the rabbit hole.

File: frontend/Content/Config.tsx [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	public name: string;
+
+	constructor(name: string) {
+		this.name = name;
+	}
+}
+// Creation of mock data, to simulate a list
+let mockData: Game[] = [];
+for (let i = 0; i < 11; i++) {
+	let newGame: Game = new Game('Super Metroid');
```
Comment (EvidentlyCube) [inline frontend/Content/Config.tsx]: So many Super Metroids 🤩

File: frontend/Components/Header.tsx [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import * as React from 'react';
+import {
+	Link,
+} from 'react-router-dom';
+export interface HeaderParams{
+	pageName: string;
+}
+export class MainHeader extends React.Component<HeaderParams> {
```
Comment (EvidentlyCube) [inline frontend/Components/Header.tsx]: Rename it to `Header` or the file to `MainHeader` - the file name should match the primary export of the file (if there is one)

File: frontend/Content/Config.tsx [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+						<p>There are a lot of spoilers in here for games you may not have played.
+                            Please select the games which youd like to see.</p>
+					</article>
+					<form action="#" method="post">
+						<button>Select All</button>
+						<button>Unselect All</button>
+						<div>
+							<input type="text" name="game" placeholder="Find a game"/>
+						</div>
+						{gameList(mockData)}
```
Comment (EvidentlyCube) [inline frontend/Content/Config.tsx]: I know it's not here to stay, but turn `gameList` into an actual component (you can pass complex data in props). A component must start with an uppercase letter btw!

File: frontend/index.tsx [hunk 8]
```diff
@@ -1,9 +1,40 @@
...
+import {Config} from './Content/Config';
 
 ReactDOM.render(
-	<Hello compiler="TypeScript" framework="React" />,
-	document.getElementById('example'),
+	<>
+		<Router>
+			<Switch>
+				<Route path="/" >
+					<Route exact path="/" component={Home}>
```
Comment (EvidentlyCube) [inline frontend/index.tsx]: Can't you replace the three lines per declaration with self closing tags?

File: frontend/index.tsx [hunk 9]
```diff
@@ -1,9 +1,40 @@
...
+					<Route path="/abilities">
+						{/* <Abilities /> */}
+					</Route>
+					<Route path="/config" component={Config}>
+						{/* <Config /> */}
+					</Route>
+				</Route>
+			</Switch>
+		</Router>
+		{/* <Hello compiler="TypeScript" framework="React" /> */}
```
Comment (EvidentlyCube) [inline frontend/index.tsx]: Kill it with fire!

File: frontend/Components/Header.tsx [hunk 10]
```diff
@@ -2,23 +2,18 @@ import * as React from 'react';
...
-}
-export class MainHeader extends React.Component<HeaderParams> {
-	private readonly pageName: string;
 
-	constructor(params: HeaderParams) {
-		super(params);
-		this.pageName = params.pageName;
-	}
+export class Header extends React.Component {
+	public state ={
```
Comment (EvidentlyCube) [inline frontend/Components/Header.tsx]: I am susprised ESLint didn't catch the lack of space between `=` and `{`.

If you want to declare `state` as a property I'd prefer it to be private, because it should never be accessed from the outside. I personally prefer to set the value of properties in the constructor rather than in the fields, because of two reasons - I want to have only one place to look at when determining default values and consistency.
You may have a situation where some of the fields always start with some constant default value, while others are set from properties. If you did it the way you did it here, you'd have to look both at the field declarations and constructor to know what value is where. But if everything is the constructor there's only one place to look.
Which segways into the consistency. if you always set the initial values in the constructor you never have to stop and think where to look for those default, because you **know** they're in the constructor. It's one of those small things that allows you to hold more things in your head at once.

Granted it's very arguable. I know some experienced developer who prefer various styles, but that's my personal favorite that I can actually give strong argumentation for :).

In other words, what I suggest:
 1. Remove the declaration of `state` field, it's not needed because it is already declared in `React.Component`
 2. Set the default value in the constructor.

File: frontend/Components/Header.tsx [hunk 11]
```diff
@@ -2,23 +2,18 @@ import * as React from 'react';
...
-	pageName: string;
-}
-export class MainHeader extends React.Component<HeaderParams> {
-	private readonly pageName: string;
 
-	constructor(params: HeaderParams) {
-		super(params);
-		this.pageName = params.pageName;
-	}
+export class Header extends React.Component {
```
Comment (EvidentlyCube) [inline frontend/Components/Header.tsx]: If you take a look at the definitions for `React.Component` you'll notice it has two generics, `P` and `S`:

```
class Component<P, S> {
```

`P` is for declaring the type of properties, and `S` is for declaring the type of the state. What I am implying here is you should add an interface for the state and pass it into the generic so that Typescript can do its job.

File: frontend/Views/Config.tsx [hunk 12]
```diff
@@ -15,22 +14,25 @@ for (let i = 0; i < 11; i++) {
...
-const gameList = function(gameData: Game[]): JSX.Element[] {
-	let result = [];
-	for (let game of gameData) {
-		result.push(<>
-			<GamesCheck game={game.name}/>
-		</>);
+interface GameListProps{
+	games: Game[];
+}
+class GameList extends React.Component<GameListProps> {
```
Comment (EvidentlyCube) [inline frontend/Views/Config.tsx]: You may want to change this to a functional component, since it has no state.

```
const GameList : React.FC<GameListProps> = (props) => {
```

Comment (EvidentlyCube) [reply frontend/Views/Config.tsx]: btw, generally always go for functional components rather than class-based components, unless you need state - in which case you have the option of turning your functinal component into a class, or trying hooks (but let's skip hooks for now).
